ofronds_internals/exercise.ml

```diff
@@ -0,0 +1,91 @@
+open! Import
+open Result.Syntax
+
+let invalid_metadata fmt =
+  let exception Invalid_metadata of string in
+  Fmt.kstr (fun s -> raise (Invalid_metadata s)) fmt
+
+type t = { name : string; path : string (* Relative to the metadata file *) }
+[@@deriving sexp]
+
+let name t = t.name
+let pp_path = Fmt.using (fun { path; _ } -> path) Fmt.string
+
+let of_stanza : Sexp.t -> t = function
+  | List (Atom "exercise" :: fields) -> t_of_sexp (List fields)
+  | sexp -> invalid_metadata "unsupported exercise format: %a" Sexp.pp sexp
+
+let with_open_in p f =
+  let+ input_channel = try Ok (open_in p) with _ -> Error `File_not_found in
+  let res = f input_channel in
+  close_in input_channel;
+  res
+
+let compile ex =
+  let ( / ) = Filename.concat in
+  let*! cmd =
+    let exe =
+      Filename.current_dir_name
+      / "exercises"
+      / (Filename.chop_extension ex.path ^ ".exe")
+    in
+    OS.Cmd.must_exist Cmd.(v "dune" % "exec" % "--force" % "--no-buffer" % exe)
```

Comment:
> Why is it better / necessary to disable buffering here?

I honestly don't know the details, but I ended up needing to do this in order to get coloured output. Dune's buffering mechanism seems to interfere with OCaml's heuristic about whether it's attached to an ANSI-compatible terminal. In theory, I think setting `OCAML_COLOR=always` should fix that, but when I was experimenting with it it didn't.
 
> It seems that you're discarding the option of having exercises of libraries instead of executables? Is that right? If so, why? I was thinking that when introducing dune-files and abstract types, it might be helpful to create libraries instead of executables. In that case, we might want to add to the metadata what we want to create and make the command here depend on that. What do you think about that, @gs0510 and @CraigFe? (I'm not asking to change that, just curious what you folks think)

That is right, and indeed this going to be an overly-restrictive assumption when teaching about Dune itself. I hadn't planned out a way to deal with that, but adding some sort of command information to the metadata sounds like a good idea.

Reply:
Re. 2, I think we can change this to rely on having libraries instead of packages. One of the ideas that @pitag-ha and I had was having dune files for exercises in `ofronds_internals` which are sym-linked.